src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java

```diff
@@ -0,0 +1,440 @@
+            estimatedParquetSize += 8;
+            break;
+          case FLOAT:
+            float floatVal =
+                DataValidationUtil.validateAndParseIcebergNumber(
+                    type.getName(), value, insertRowsCurrIndex, Float.class);
+            value = floatVal;
+            stats.addRealValue((double) floatVal);
+            estimatedParquetSize += 4;
+            break;
+          case DOUBLE:
+            double doubleVal =
+                DataValidationUtil.validateAndParseIcebergNumber(
+                    type.getName(), value, insertRowsCurrIndex, Double.class);
+            value = doubleVal;
+            stats.addRealValue(doubleVal);
+            estimatedParquetSize += 8;
+            break;
+          case BINARY:
+            byte[] byteVal = getBinaryValue(value, primitiveType, stats, insertRowsCurrIndex);
+            value = byteVal;
+            estimatedParquetSize +=
+                ParquetBufferValue.BYTE_ARRAY_LENGTH_ENCODING_BYTE_LEN + byteVal.length;
+            break;
+          case FIXED_LEN_BYTE_ARRAY:
+            byte[] fixedLenByteArrayVal =
+                getFixedLenByteArrayValue(value, primitiveType, stats, insertRowsCurrIndex);
+            value = fixedLenByteArrayVal;
+            estimatedParquetSize +=
+                ParquetBufferValue.BYTE_ARRAY_LENGTH_ENCODING_BYTE_LEN
+                    + fixedLenByteArrayVal.length;
+            break;
+          default:
+            throw new SFException(
+                ErrorCode.UNKNOWN_DATA_TYPE,
+                type.getLogicalTypeAnnotation(),
+                primitiveType.getPrimitiveTypeName());
+        }
+      } else {
+        return getGroupValue(
+            value, type.asGroupType(), stats, defaultTimezone, insertRowsCurrIndex);
+      }
+    }
+
+    if (value == null) {
+      if (!type.isRepetition(Repetition.REQUIRED)) {
+        throw new SFException(
+            ErrorCode.INVALID_FORMAT_ROW, type.getName(), "Passed null to non nullable field");
+      }
+      stats.incCurrentNullCount();
+    }
+
+    return new ParquetBufferValue(value, estimatedParquetSize);
+  }
+
+  /**
+   * Parses an int32 value based on Parquet logical type.
+   *
+   * @param value column value provided by user in a row
+   * @param type Parquet column type
+   * @param insertRowsCurrIndex Used for logging the row of index given in insertRows API
+   * @return parsed int32 value
+   */
+  private static int getInt32Value(
+      Object value, PrimitiveType type, final long insertRowsCurrIndex) {
+    LogicalTypeAnnotation logicalTypeAnnotation = type.getLogicalTypeAnnotation();
+    if (logicalTypeAnnotation == null) {
+      return DataValidationUtil.validateAndParseIcebergNumber(
+          type.getName(), value, insertRowsCurrIndex, Integer.class);
+    }
+    if (logicalTypeAnnotation instanceof DecimalLogicalTypeAnnotation) {
+      return getDecimalValue(value, type, insertRowsCurrIndex).unscaledValue().intValue();
```

Comment:
why not use columnMetadata.scale ?

Reply:
also, columnMetadata.precision isn't needed here?

Reply:
yep, checked documentation we do support custom precision and scale to be specified for managed iceberg tables' decimal fields. See here: https://docs.snowflake.com/en/user-guide/tables-iceberg-data-types

Reply:
also applies to long decimal..

Reply:
The purpose of this function is to parse java type to parquet bytes. The actual bytes of a parquet decimal column is an unscaled integer. The scanner later use the precision and scale in schema to infer the correct information.